recent topics: Highlight the filter term when selecting input after navigation.

legengen opened the pull request:
**Summary of Changes:**
Add logic to the show function: When the filter box is not empty, set should_select_search_text to true.
Add logic in setDefaultFocus: Determine whether to perform the select or focus action based on should_select_search_text.

**Reason for correctness:**
Only pass the filter state, and leave it to other logic to decide whether to highlight the filter box.

**Focus Behavior:**
When the focus is on topic_row, switching back to "Recent conversations" will not change the focus to the search box.
When there is no focus on the screen or the focus is on the search box, switching back to "Recent conversations" will change the focus to the search box.
When the user is typing, the focus will not be changed.

<!-- Describe your pull request here.-->

**Fixes:** https://github.com/zulip/zulip/issues/17941<!-- Issue link, or clear description.-->

**How changes were tested:**
manual testing

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![gif1](https://github.com/user-attachments/assets/21e11de6-a6f5-4aea-9833-f7673aa61e5d)
![gif2](https://github.com/user-attachments/assets/fb5c3556-7750-457b-b166-4efdc285c94f)

<details>

<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [X] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [X] Highlights technical choices and bugs encountered.
- [X] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [X] Each commit is a coherent idea.
- [X] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [X] Visual appearance of the changes.
- [X] Responsiveness and internationalization.
- [X] Strings and tooltips.
- [X] End-to-end functionality of buttons, interactions and flows.
- [X] Corner cases, error conditions, and easily imagined bugs.
</details>

legengen committed:
recent topics: Highlight the filter term when selecting input after navigation.
Fixes zulip#17941
Add logic to the show function: When the filter box is not empty, set should_select_search_text to true.
Add logic in setDefaultFocus: Determine whether to perform the select or focus action based on should_select_search_text.

alya commented:
Hm, I'm confused -- I'm not seeing the filter term getting highlighted in those screen captures?

legengen commented:
@alya Sorry, my previous screenshot was not captured properly. Here are the new screenshots along with explanations:

1. When switching from other interfaces to the Recent Conversation view, the filtered keywords are highlighted.
![gif3](https://github.com/user-attachments/assets/faa40772-d6ba-497b-aa56-22bd72775187)

2. When a topic_row is already selected, switching to Recent Conversation does not highlight the filtered keywords to avoid disturbing the user experience.
![gif4](https://github.com/user-attachments/assets/2a1a0c11-bd0e-4748-8489-6fbdb7a10eb0)

From the discussions in the PR for the previous issue, I learned that the functionality shown in Screenshot 2 needs to be implemented. Under certain conditions, the filtered keywords will not be highlighted, which might be why you were confused about the lack of highlighting.

alya commented:
Thanks! @amanagr Could you please take a look? I have't tested.

amanagr commented on web/src/recent_view_ui.ts:
```diff
@@ -1411,6 +1412,7 @@ export function update_recent_view_rendered_time(): void {
     }
 }
 
+let should_select_search_text = false;
```
Should be reset to `false` once recent view has finished rendering?

legengen committed:
inbox_ui: Highlight the filter term when navigating in inbox and channel view.
Implement highlight-on-navigation behavior for the Inbox search box and the topic list filter in channel-view.

legengen commented:
I think I have completed the same work in inbox-view and channel-view, here are the screenshots.

inbox-view case: When switching from other pages to inbox-view, highlight it, but when returning after selecting a topic, do not highlight to avoid disturbing UX.
![inbox-view-case](https://github.com/user-attachments/assets/f548e790-883a-4914-a388-aad53fe42cbe)
channel-view case: Same as above.
![channel-view-case](https://github.com/user-attachments/assets/68e26e07-f6c6-474d-a810-64937459b2cf)

legengen commented on web/src/recent_view_ui.ts:
```diff
@@ -152,8 +152,9 @@ export function is_in_focus(): boolean {
 export function set_default_focus(): void {
     // If at any point we are confused about the currently
     // focused element, we switch focus to search.
+    const focusAction = should_select_search_text ? "select" : "focus";
```
I’ve rebased the branch and fixup-merged the commits into 3cc734e. Please review.

legengen commented on web/src/inbox_ui.ts:
```diff
@@ -372,6 +377,10 @@ export function show(filter?: Filter): void {
 
+    if (search_keyword !== "") {
+        should_select_inbox_search = true;
+    }
```
```diff
    inbox_util.set_filter(filter);
+   if (search_keyword !== "") {
+        should_select_inbox_search = true;
+   }
    if (inbox_util.is_channel_view()) {
        restore_channel_view_state();
        views_util.show({
```
I’ve extracted the should_select_inbox_search = true logic to above the if (inbox_util.is_channel_view()) block. Manual testing shows that the behavior remains the same. Please review.

amanagr commented on web/src/inbox_ui.ts:
```diff
@@ -1510,7 +1520,12 @@ export function revive_current_focus(): void {
+            return;
+        }
+        select_inbox_search();
+        should_select_inbox_search = false;
```
Please let me know once this is ready for review.

legengen commented:
Hi! Due to time constraints, I won’t be able to continue maintaining this PR, so I’m stepping away. The implementation is in a good state and should be straightforward to finish.

**What’s done**
- Implemented highlight/select behavior for the filter term when navigating back to Recent conversations (via should_select_search_text + setDefaultFocus).
- Extended the same behavior to Inbox view search and Channel view topic list filter (screenshots/commits included).
- Started a CZO #design thread to settle the exact UX across Recent/Inbox/Channel.

**What’s left**
- Address remaining review feedback in inbox_ui.ts regarding focus vs select logic (see the comment about focus_current_id vs focus_inbox_search and the suggested extra condition).
- Design decision pending (CZO #design): there isn’t a final decision yet on the exact behavior across Recent/Inbox/Channel. The next maintainer should continue that discussion and then align the implementation accordingly.

**Handoff**
Feel free to take it over (either push commits to this branch / or open a follow-up PR). Please proceed without waiting for me.
